### ISSUE TITLE
Add Serde (de)serialization

### ISSUE THREAD

**Issue by pimeys** [tiberius-rs/tiberius#115]
We could maybe have a nice interface for people using serde structs in their code. Let's see a few examples:

```rust
#[derive(Deserialize)]
struct Cat {
    id: i64,
    name: String
}
```

Now using the struct, we could deserialize rows directly to them in an efficient manner:

```rust
let mut client = tiberius::Client::connect(config, tcp.compat_write()).await?;
let stream = client.query("SELECT id, name FROM cats", &[&1i32, &2i32, &3i32]).await?;
let rows = streams.into_first_result().await?;

let cats = stream
    .by_ref()
    .map_ok(|row| row.deserialize::<Cat>())
    .try_collect()
    .await?;
```

The question is, should Tiberius do this, or should it be let for higher-level crates, such as Diesel?

**Comment by yatesco** [comment]
On first sight this looks appealing but in practice you're concreting column names in the database with the names in the JSON that is inevitably exported. In real life it would be more like:

```rust
#[derive(Deserialize)]
struct Cat {
    id: i64,
    #[serde(rename="someField")] // for JSON, but do I want "someField" in my database
    some_field: String
}
```

I guess you could go down the whole "value object" route but blergh.

BTW, I tend not to do either of these - I tend to store the serialised JSON in the database and use `json_value` etc. to expose indexed fields, so I'm not really the target audience for this.

I do see the appeal for this during onboarding and for more isolated/smaller projects, but decades of pain have taught me ORMapping almost always fail :-)

**Comment by pimeys** [comment]
We don't yet have `FromRow` trait, only [FromSql](https://docs.rs/tiberius/0.5.8/tiberius/trait.FromSql.html) that's per-column. I wouldn't say no to a PR to implement that though, that would be quite useful for many users and not hard at all to write.

And, I agree with you @yatesco, I don't really feel the need for serde support in this crate.

**Comment by mlcruz** [comment]
I have been working for a while on something similar

Right now i have a [very simple working example](https://github.com/mobiltracker/tiberius-derive/blob/master/src/lib.rs) for a derive macro that implements  TryFrom<tiberius::Row> for some struct. I'm quite new to proc-macros, so its not that great, but works well enough for our internal usage and its better than manually getting each field.

```rust
#[derive(TiberiusRow)]
struct Foobar {
    foo: String,
    bar: bool,
    zar: Option<String>,
    xar: Option<bool>,
}
```

Expands to:

```rust
impl TryFrom<tiberius::Row> for Foobar {
    type Error = tiberius::error::Error;
    fn try_from(__row: tiberius::Row) -> Result<Self, Self::Error> {
        Ok(Self {
            foo: {
                __row
                    .try_get::<&str, &str>("foo")?
                    .expect(&{
                        let res = ::alloc::fmt::format(::core::fmt::Arguments::new_v1(
                            &["Failed to get field "],
                            &match (&"foo",) {
                                (arg0,) => [::core::fmt::ArgumentV1::new(
                                    arg0,
                                    ::core::fmt::Display::fmt,
                                )],
                            },
                        ));
                        res
                    })
                    .to_owned()
            },
            bar: {
                __row.try_get("bar")?.expect(&{
                    let res = ::alloc::fmt::format(::core::fmt::Arguments::new_v1(
                        &["Failed to get field "],
                        &match (&"bar",) {
                            (arg0,) => [::core::fmt::ArgumentV1::new(
                                arg0,
                                ::core::fmt::Display::fmt,
                            )],
                        },
                    ));
                    res
                })
            },
            zar: { __row.try_get::<&str, &str>("zar")?.map(|f| f.to_owned()) },
            xar: { __row.try_get("xar")? },
        })
    }
}
```

I wouldn't mind improving on the implementation and submitting a PR. but i'm not sure if a derive macro is the right way to do this.

**Comment by tomhoule** [comment]
Hi @micruz! I am working with database drivers all the times and have written a lot of Rust macro code. I implemented [something similar with serde in quaint](https://github.com/prisma/quaint/blob/master/src/serde.rs) and with hindsight, I definitely think a custom macro the way you are doing it is better: it generates less code and it can live as a separate crate.

One design consideration is whether you get the columns by name, like in your example, or by position (forcing the struct fields to be in the same order as the values in your `SELECT` clause). By position adds this constraint, but it's faster. The diesel ORM does it that way for example.

I think something like that as a separate crate or an optional feature in tiberius would be great :)

**Comment by pimeys** [comment]
See also how ScyllaDB handles the de-serialization. Not saying which idea is the best, but it's interesting to see people taking different approaches to this problem:

https://github.com/scylladb/scylla-rust-driver/blob/main/examples/select-paging.rs#L36

**Comment by pimeys** [comment]
Also, what I'd like to see is that if we take the proc macro route, that we could instead of copying, move the data from the row to the struct. It can really slow up the execution if we clone in this macro, without the developer noticing...

**Comment by tomhoule** [comment]
Yep, or have the struct hold references to the contents of the row. `struct MyCustomRow<'a> { ... }`.

**Comment by jsouto18** [comment]
Is anyone working on this feature?

**Comment by jsouto18** [comment]
@pimeys What do you think of the implementation done on https://github.com/mobiltracker/tiberius-derive?

**Comment by MattJackson** [comment]
#442 (in review) adds optional serde `Serialize`/`Deserialize` on the result types (`Row`/`ColumnData`). The direct `row.deserialize::<MyStruct>()` mapping sketched here isn't part of that, so leaving open to track it — including the question of whether it belongs here vs a higher-level crate.

**Comment by MattJackson** [comment]
Fixed in #442, shipped in **v0.13.0** (https://github.com/tiberius-rs/tiberius/releases/tag/v0.13.0). Closing.
